### DISCUSSION TITLE
Does Equinox collect metrics around ClassLoading?

### DISCUSSION THREAD

**Discussion by guw** [eclipse-equinox/equinox#633] [Q&A] [answered]
Is there a way to gather class loading statistics from Equinox?

I am looking for:
* number of classes loaded (since the framework was launches)
* total time spent loading classes (since the framework was launches)

The [documentation mentions such an adapter](https://github.com/eclipse-equinox/equinox/blob/master/docs/Adaptor_Hooks.md#StatsManager) but it looks like the code [has been deleted in Luna](https://bugs.eclipse.org/bugs/show_bug.cgi?id=412502#c1)?

The `ClassLoaderStatsHook` mentioned in the documentation also seems to no longer exist. I was able to find [StatsManager](https://github.com/eclipse-equinox/equinox/blob/b5de515e51604f756cab3b6cd2188574b8aecc3f/bundles/org.eclipse.osgi/eclipseAdaptor/src/org/eclipse/core/runtime/internal/stats/StatsManager.java) in Git history. Got the functionality purged completely without replacement?

**Comment by tjwatson** [comment, marked as answer]
This got removed a very long time ago when the adapter hooks got removed in 8d9916919400ae6071271bc23f341bd08e80f802

There has not been any effort to build something back into the framework.  Although I would advocate it not be built into the framework and instead provided by a system bundle fragment instead.

**Reply by guw** [reply]
Thanks @tjwatson. Is there any guidance on how to create a system bundle fragment that plugs into Equinox automatically?

**Reply by tjwatson** [reply]
Are you sure you are observing performance issues with class loading?  With Eclipse the typical slow down during class loading would be the impact of lazy activation which causes a bundle to be activated upon first class load.  Maybe enable one of the following Equinox trace options to see how long it is taking to start a bundle:

```
# Prints out timing information for loading and calling BundleActivator.start
org.eclipse.osgi/debug/bundleTime=false
# Prints out total time for starting a bundle which includes:
# - loading and calling BundleActivator.start
# - other time spent resolving bundles and firing bundle events to listeners
org.eclipse.osgi/debug/bundleStartTime=false
```

To get some stats around class loading you would need to implement a `org.eclipse.osgi.internal.hookregistry.ClassLoaderHook`.  But to do that you cannot simply be a system bundle fragment.  You would need to be an Equinox specific extension similar to `org.eclipse.equinox.weaving.hooks.WeavingHook`

**Reply by guw** [reply]
Thanks @tjwatson. 

For a bit more context, it's not an Eclipse application. It's more like an app server application where a single web app gets deployed as bundle. Web apps can be a composition a bundle and fragments with thousands of entries in `Bundle-Classpath` header. We recently change the granularity of jars (significantly), which led to a significant increase in start time. That's what I am investigating.

For example, one discover we made during the exercise was that `external:/absolute/path/to/folder` performs much worse then `relative/path/to/folder`. Have not profiled into details why, though.

**Reply by tjwatson** [reply]
When you say 1000s of entries, you mean a comma separated list of thousands of JARs?  If so then the osgi.bundlefile.limit will likely have a big impact with its default of 100.  It would mean for each class load on a bundle it may need to open/close a 1000 bundle files to keep the limit of 100 concurrently opened bundle files JARs.  You may want to try disabling that or bump it up significantly.  You disable it by setting the system property (config.ini) value `osgi.bundlefile.limit=0`.

**Reply by guw** [reply]
Confirming. We use `osgi.bundlefile.limit=0`.

**Reply by tjwatson** [reply]
OK, then not sure what the issue is.  The same scenario in https://www.eclipse.org/forums/index.php?t=msg&th=1111465&goto=1854519& raising the limit helped out.

**Reply by guw** [reply]
I am not suspecting there is an issue in Equionx. The reason I want the stats is to have more data for comparison and analysis. There are "_interesting_" things happening inside the webapp, some of which may be impacted directly by classpath size. Thus, the data is more helpful to rule out Equinox as an issue in this case.

**Comment by akurtakov** [comment]
You may want to give a try to systemtap-java https://developers.redhat.com/blog/2014/01/10/probing-java-w-systemtap#basic_usage
